### PR TITLE
Fixes # 10592. docker save by image id generates repositories JSON

### LINKED ISSUE #10592: Docker Save Command Issue
Description of problem:

When saving a repo, if you specify the image by ID, then the FQIN is discarded.  On load, the result is a '<none>' name '<none>' tag repo which isn't very user-friendly.  However, saving by FQIN does preserve the repo. name/tag, such that loading the tarball restores them.

`docker version`:

Client version: 1.4.1-dev
Client API version: 1.17
Go version (client): go1.4.1
Git commit (client): e1b875c
OS/Arch (client): linux/amd64
Server version: 1.4.1-dev
Server API version: 1.17
Go version (server): go1.4.1
Git commit (server): e1b875c

`docker info`:

Containers: 93
Images: 59
Storage Driver: devicemapper
 Pool Name: docker-253:2-2766409-pool
 Pool Blocksize: 65.54 kB
 Backing Filesystem: extfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 5.006 GB
 Data Space Total: 107.4 GB
 Metadata Space Used: 10.06 MB
 Metadata Space Total: 2.147 GB
 Udev Sync Supported: true
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.90 (2014-09-01)
Execution Driver: native-0.2
Kernel Version: 3.17.7-300.fc21.x86_64
Operating System: Fedora 21 (Twenty One)
CPUs: 4
Total Memory: 11.43 GiB
Name: dhcp-10-19-62-194.boston.devel.redhat.com
ID: R54O:3ZKE:PREZ:FO6U:C36U:TN45:4VVI:B5O3:OZXT:O5QG:OWTE:47IR
Debug mode (server): true
Debug mode (client): false
Fds: 12
Goroutines: 17
EventsListeners: 0
Init SHA1: 1b7372ffe99bf1de06bef70cd8f2108335de676a
Init Path: /home/shishir/docker/src/github.com/docker/docker/bundles/1.4.1-dev/dynbinary/dockerinit
Docker Root Dir: /var/lib/docker

`uname -a`:
Linux dhcp-10-19-62-194.boston.devel.redhat.com 3.17.7-300.fc21.x86_64 #1 SMP Wed Dec 17 03:08:44 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

Environment details (AWS, VirtualBox, physical, etc.):
Physical, Fedora 21

How reproducible:
100%

Steps to Reproduce:
1. docker save -o <image name>.tar <image name>
2. docker save -o <image ID>.tar <image ID>
3. docker rmi <image name>
4. docker load -i <image ID>.tar
5. docker images
6. docker load -i <image name>.tar
7. docker images

Actual Results:

[root@dockertest ~]# docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
docker_testimage    test_tag            9e4968c9710f        18 minutes ago      499.5 MB
[root@dockertest ~]# docker save -o docker_testimage:test_tag.tar docker_testimage:test_tag
[root@dockertest ~]# docker save -o 9e4968c9710f.tar 9e4968c9710f
[root@dockertest ~]# docker rmi docker_testimage:test_tag
Untagged: docker_testimage:test_tag
Deleted: 9e4968c9710f332e416307a1ac25cc80c5c2a85271234e38022ce300fa966c34
...cut...
Deleted: bef54b8f8a2fdd221734f1da404d4c0a7d07ee9169b1443a338ab54236c8c91a
[root@dockertest ~]# docker load -i 9e4968c9710f.tar
[root@dockertest ~]# docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
<none>              <none>              9e4968c9710f        23 minutes ago      499.5 MB
[root@dockertest ~]# docker load -i docker_testimage\:test_tag.tar 
[root@dockertest ~]# docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
docker_testimage    test_tag            9e4968c9710f        23 minutes ago      499.5 MB

Expected Results:

[root@dockertest ~]# docker load -i 9e4968c9710f.tar
[root@dockertest ~]# docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
docker_testimage    test_tag            9e4968c9710f        23 minutes ago      499.5 MB
[root@dockertest ~]# docker load -i docker_testimage\:test_tag.tar 
[root@dockertest ~]# docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
docker_testimage    test_tag            9e4968c9710f        23 minutes ago      499.5 MB

Additional info:
In the case of saving by ID, all the FQIN's which point at that ID should be recorded.  On load, only the non-conflicting names should be restored, preferably issuing a warning about any name conflicts

### PR DESCRIPTION
**Pull request by shishir-a412ed**
Fixes #10592 
Signed-off-by: Shishir Mahajan shishir.mahajan@redhat.com

### CODE REVIEW

**Comment by icecrime** [comment]
What about https://github.com/docker/docker/issues/10592#issuecomment-125330209?

**Comment by shishir-a412ed** [comment]
ping @crosbymichael 
This is inline with what we discussed on irc.

Docker save by `IMG_ID` generates the repositories JSON, which preserves all the associated names and tags with that `IMG_ID`. No change to docker load. It is able to handle this smoothly and load all the tags.

Shishir

**Comment by shishir-a412ed** [comment]
@crosbymichael Any updates on this one ?

**Comment by shishir-a412ed** [comment]
@crosbymichael seems to be pretty occupied :)

@icecrime can someone else review this PR ?
ping @LK4D4 @tiborvass 

Shishir

**Comment by thaJeztah** [comment]
ping @icecrime @crosbymichael for review, pls.

**Comment by calavera** [comment]
I'm not completely sure what this does. Can we have some test cases to see it working?

**Comment by shishir-a412ed** [comment]
@calavera This PR fixes this.

https://github.com/docker/docker/issues/10592#issue-56693115

https://github.com/docker/docker/issues/10592#issuecomment-128024349

**Comment by calavera** [comment]
@shishir-a412ed tests are still very appreciated to make sure we don't introduce regressions in the future.

**Comment by shishir-a412ed** [comment]
@calavera I was not trying to escape the tests :)
Just wanted to make sure that you understand the issue, before I invest more time on it.

Shishir

**Comment by shishir-a412ed** [comment]
@calavera test added.

**Comment by shishir-a412ed** [comment]
@calavera @thaJeztah 
Any updates on this one ?

File: integration-cli/docker_cli_save_load_test.go [hunk 1]
```diff
@@ -97,6 +97,56 @@ func (s *DockerSuite) TestSaveSingleTag(c *check.C) {
...
+	// This load should populate the non-conflicting tags back.
+	// After this load, the image should have all 3 tags.
+	loadCmd := exec.Command(dockerBinary, "load", "-i", filepath.Join(tmpDir, "emptyfs.tar"))
+	if _, err := loadCmd.CombinedOutput(); err != nil {
+		c.Errorf("failed to run docker load command: %s", err)
+	}
+
+	out, _, err = runCommandPipelineWithOutput(
+		exec.Command(dockerBinary, "images"),
+		exec.Command("grep", "-E", "(save-load-multiple-tags-one|save-load-multiple-tags-two)"),
```
Comment (calavera) [inline integration-cli/docker_cli_save_load_test.go]: I'm not sure this pipeline is very portable. I think you can just call docker images and check if both tags are in the output, can't you?

File: graph/export.go [hunk 2]
```diff
@@ -61,6 +62,20 @@ func (s *TagStore) ImageExport(names []string, outStream io.Writer) error {
...
+					//This must be an ID. Here repoName is an ID.
+					nonTruncID, err := s.graph.GetTruncIndex().Get(repoName)
+					if err != nil {
+						return err
+					}
+					reverseLookupMap := s.ByID()
+					for id, repos := range reverseLookupMap {
+						if id == nonTruncID {
+							for _, repo := range repos {
+								addKey(repo[:strings.Index(repo, ":")], repo[strings.Index(repo, ":")+1:], id)
```
Comment (calavera) [inline graph/export.go]: can you extract `strings.Index...` and assign it to a variable? it's kinda hard to read and duplicated

File: integration-cli/docker_cli_save_load_test.go (hunk 1 above)
Comment (shishir-a412ed) [inline integration-cli/docker_cli_save_load_test.go]: Fixed.

File: graph/export.go (hunk 2 above)
Comment (shishir-a412ed) [inline graph/export.go]: Fixed.

**Comment by shishir-a412ed** [comment]
@calavera Please check now.

File: graph/export.go [hunk 3]
```diff
@@ -61,6 +62,21 @@ func (s *TagStore) ImageExport(names []string, outStream io.Writer) error {
...
+					//This must be an ID. Here repoName is an ID.
+					nonTruncID, err := s.graph.GetTruncIndex().Get(repoName)
+					if err != nil {
+						return err
+					}
+					reverseLookupMap := s.ByID()
+					for id, repos := range reverseLookupMap {
+						if id == nonTruncID {
+							for _, repo := range repos {
+								index := strings.Index(repo, ":")
```
Comment (calavera) [inline graph/export.go]: oh, I'm very sorry, I should have thought about this before, isn't this like doing?

```go
repoName, repoTag := strings.SplitN(repo, ":", 2)
addKey(repoName, repoTag, id)
```

**Comment by calavera** [comment]
one minor thing, but otherwise LGTM.

**Comment by shishir-a412ed** [comment]
@calavera 
No worries :) .. check now.

**Comment by calavera** [comment]
nice, thank you!

**Comment by shishir-a412ed** [comment]
ping @tiborvass @crosbymichael

File: graph/export.go [hunk 4]
```diff
@@ -61,6 +62,21 @@ func (s *TagStore) ImageExport(names []string, outStream io.Writer) error {
...
 					addKey(repoName, repoTag, img.ID)
+				} else {
+					//This must be an ID. Here repoName is an ID.
+					nonTruncID, err := s.graph.GetTruncIndex().Get(repoName)
+					if err != nil {
+						return err
+					}
+					reverseLookupMap := s.ByID()
+					for id, repos := range reverseLookupMap {
+						if id == nonTruncID {
```
Comment (tiborvass) [inline graph/export.go]: `reverseLookupMap[id]` is O(1), no need to loop through the map.

**Comment by thaJeztah** [comment]
@tiborvass I see this PR was updated; PTAL

**Comment by shishir-a412ed** [comment]
@tiborvass Any updates on this one ?

N thanks for the catch on map lookup :) Constant time is always better than linear time complexity.

File: integration-cli/docker_cli_save_load_test.go [hunk 5]
```diff
@@ -97,6 +97,60 @@ func (s *DockerSuite) TestSaveSingleTag(c *check.C) {
 	}
 }
 
+// TestSaveAndLoadMultipleTags test fix for Issue #10592.
+func (s *DockerSuite) TestSaveAndLoadMultipleTags(c *check.C) {
+	repoNameOne := "save-load-multiple-tags-one"
+	repoNameTwo := "save-load-multiple-tags-two"
+
+	// Tag emptyfs:latest with two more tags.
+	dockerCmd(c, "tag", "emptyfs:latest", fmt.Sprintf("%v:latest", repoNameOne))
```
Comment (cpuguy83) [inline integration-cli/docker_cli_save_load_test.go]: Should just be `%s`, but also `repoNameOne+":latest"` is also better.

File: integration-cli/docker_cli_save_load_test.go [hunk 6]
```diff
@@ -97,6 +97,60 @@ func (s *DockerSuite) TestSaveSingleTag(c *check.C) {
...
+	dockerCmd(c, "tag", "emptyfs:latest", fmt.Sprintf("%v:latest", repoNameTwo))
+
+	tmpDir, err := ioutil.TempDir("", "save-load-multiple-tags")
+	if err != nil {
```
Comment (cpuguy83) [inline integration-cli/docker_cli_save_load_test.go]: `c.Assert(err, check.IsNil)`

File: integration-cli/docker_cli_save_load_test.go [hunk 7]
```diff
@@ -97,6 +97,60 @@ func (s *DockerSuite) TestSaveSingleTag(c *check.C) {
...
+		c.Errorf("failed to create temporary directory: %s", err)
+	}
+
+	out, _ := dockerCmd(c, "images", "-q", repoNameOne)
+	cleanedShortImageID := strings.TrimSpace(out)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_save_load_test.go]: How about just `shortID`?

File: integration-cli/docker_cli_save_load_test.go [hunk 8]
```diff
@@ -97,6 +97,60 @@ func (s *DockerSuite) TestSaveSingleTag(c *check.C) {
...
+
+	// Save the image into a tarball. This should preserve all three tags.
+	saveCmd := exec.Command(dockerBinary, "save", cleanedShortImageID)
+	saveOut, err := saveCmd.CombinedOutput()
+	if err != nil {
```
Comment (cpuguy83) [inline integration-cli/docker_cli_save_load_test.go]: `c.Assert(err, check.IsNil, check.Commentf(string(saveOut)))`

File: integration-cli/docker_cli_save_load_test.go [hunk 9]
```diff
@@ -97,6 +97,60 @@ func (s *DockerSuite) TestSaveSingleTag(c *check.C) {
...
+		c.Errorf("failed to run docker save command: %s", err)
+	}
+	if err := ioutil.WriteFile(filepath.Join(tmpDir, "emptyfs.tar"), saveOut, 0666); err != nil {
+		c.Errorf("failed to create tar ball: %s", err)
+	}
+
+	// Untag tags: save-load-multiple-tags-one and save-load-multiple-tags-two.
+	rmiCmd := exec.Command(dockerBinary, "rmi", repoNameOne, repoNameTwo)
+	if _, err := rmiCmd.CombinedOutput(); err != nil {
+		c.Errorf("failed to run docker rmi command: %s", err)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_save_load_test.go]: Please include the output on the error message, also maybe just:
```go
out, err := rmiCmd.CombinedOutput()
c.Assert(err, check.IsNil, string(out))
```

File: integration-cli/docker_cli_save_load_test.go [hunk 10]
```diff
@@ -97,6 +97,60 @@ func (s *DockerSuite) TestSaveSingleTag(c *check.C) {
...
+
+	out, _ = dockerCmd(c, "images", "save-load-multiple-tags-one")
+
+	if !strings.Contains(out, "save-load-multiple-tags-one") {
+		c.Fatal("images should've listed save-load-multiple-tags-one")
+	}
+
+	out, _ = dockerCmd(c, "images", "save-load-multiple-tags-two")
+
+	if !strings.Contains(out, "save-load-multiple-tags-two") {
```
Comment (cpuguy83) [inline integration-cli/docker_cli_save_load_test.go]: `checker.Contains`

**Comment by shishir-a412ed** [comment]
@cpuguy83 Fixed. Please check now.

**Comment by cpuguy83** [comment]
Can you add a test verifying that an image tag that exists in the saved tarbal does not overwrite one that exists in the graph (presumably with a different ID).

**Comment by shishir-a412ed** [comment]
@cpuguy83 Do you want this to be a separate test ? or should I include this condition in the existing test `TestSaveAndLoadMultipleTags` I have added ?

**Comment by cpuguy83** [comment]
Should be ok in the same test.

**Comment by shishir-a412ed** [comment]
@cpuguy83 I checked this scenario. This seems to be a problem with docker master (load functionality).
I will give an example.

> docker images
REPOSITORY          TAG                 IMAGE ID               CREATED             VIRTUAL SIZE
yahoo                        latest              d7057cb02084        12 days ago           1.096 MB
busybox                    latest              d7057cb02084        12 days ago            1.096 MB
fedora                       latest              ded7cd95e059         4 months ago        186.5 MB

> docker save yahoo:latest > yahoo.tar
> docker rmi -f d7057cb02084

> docker images 
REPOSITORY          TAG                 IMAGE ID               CREATED             VIRTUAL SIZE
fedora                       latest                ded7cd95e059         4 months ago        186.5 MB

> docker tag fedora yahoo
REPOSITORY          TAG                 IMAGE ID               CREATED             VIRTUAL SIZE
fedora                       latest                ded7cd95e059         4 months ago        186.5 MB
yahoo                       latest                ded7cd95e059         4 months ago        186.5 MB

> docker load < yahoo.tar
The image yahoo:latest already exists, renaming the old one with ID ded7cd95e059 to empty string

I believe this is an expected behavior for one usecase where someone is trying to load a new version. The old one gets renamed to empty string. But it should check for IDs.

This is not a problem with `docker tag` command.

> docker images
REPOSITORY          TAG                 IMAGE ID               CREATED             VIRTUAL SIZE
yahoo                        latest              d7057cb02084        12 days ago           1.096 MB
busybox                    latest              d7057cb02084        12 days ago            1.096 MB
fedora                       latest              ded7cd95e059         4 months ago        186.5 MB

> docker tag fedora yahoo
Error response from daemon: Conflict: Tag latest is already set to image d7057cb020844f245031d27b76cb18af05db1cc3a96a29fa7777af75f5ac91a3, if you want to replace it, please use -f option.

If you are okay, I can open another PR to fix this problem (docker load) first. Then I can add a test here and we can land this on top of that PR.

Shishir

**Comment by cpuguy83** [comment]
@shishir-a412ed I see, no need to change the existing behavior.

**Comment by shishir-a412ed** [comment]
@cpuguy83 
I think you misunderstood my comment. This is not a change to existing behavior but a bug in the master. 
Even if the old tag is getting renamed to empty string by a tarball being loaded, it should only rename tag of its own image. Not of some other image, which it has no business with.

**Comment by shishir-a412ed** [comment]
@cpuguy83 any updates on this ?
Ping @tiborvass @calavera @thaJeztah

**Comment by thaJeztah** [comment]
Sorry for the delay, @shishir-a412ed I'll ask if someone is able to have a look soon :+1:

**Comment by rhatdan** [comment]
@tiborvass @calavera @thaJeztah @jfrazelle @crosbymichael Could someone give us feedback on this.  We need this to optimize some functionality we are writing to allow people to move from one docker back end to another.

**Comment by thaJeztah** [comment]
@rhatdan @shishir-a412ed planning to have a PR-review round this afternoon, I'll make sure this one gets added to the list there

**Comment by bfirsh** [comment]
@thaJeztah was this taken a look at in the end?

**Comment by shishir-a412ed** [comment]
@bfirsh I left a comment for @cpuguy83 .

https://github.com/docker/docker/pull/15263#issuecomment-145365510
https://github.com/docker/docker/pull/15263#issuecomment-145917418

I believe you have a bug in the master branch, but if you feel this is an existing behavior.
Then this PR has everything you need, including tests. You should be able to merge it then.

**Comment by tonistiigi** [comment]
I see this is already on code review so sorry for being late but have to mark my `-1` on the design. Image IDs and tags are different concepts and will have independent interfaces in `v1.10`. Image IDs will become content addressable and carry lot of meaning but no meaning about the tag. It will be useful to save/load plain image IDs because they refer to immutable content, compared to tags that just represent a random state in the time `docker save` was called. 

For the patch, it reintroduces the digests in the tar, resulting broken tags after loading.

**Comment by aidanhs** [comment]
@tonistiigi do you have a link where we can read more about how the interfaces are changing for v1.10?

**Comment by tonistiigi** [comment]
@aidanhs We will soon, we will not remove features if thats what you are worried about. Just big refactoring of the code.

**Comment by shishir-a412ed** [comment]
@tonistiigi
IIUC you will not be removing any features to maintain backward compatibility. At the same time, you are trying to associate imageID with immutable content and no relation with tags.

I would like to know how that would solve my problem of `loading a tarball resulting in an untagged image.`
Specifically this usecase:
https://github.com/docker/docker/issues/10592#issuecomment-128024349

Right now saving a tarball using an imageID, and then loading it somewhere else results in a none:none (untagged) image,

**Comment by tonistiigi** [comment]
@shishir-a412ed I just think it doesn't fit into the design not that we would make it impossible to solve. For your usecase, I think its wrong expectation. To me, if you save by image ID you should only get the image ID when you load. If you want to save with 3 tags you should do `docker save tag1 tag2 tag3`.

**Comment by shishir-a412ed** [comment]
@tonistiigi okay, think about a usecase where I (user) am trying to migrate my docker images from one docker instance to another. After the migration I would like to see my images at the new docker instance, the same way they were at the source. 
So in the above example if my image has N tags, I would need to do `docker save tag1 tag2 tag3 tag4 ... tagN` and create N tarballs of the same image to get my image and tags the same way they were at the source.

This is the problem, why we need this feature.
An actual production usecase would be, switching the docker storage backend from devicemapper to overlayfs. In that case, the user would like to do:

1) save docker images
2) switch docker storage
3) load docker images.

Another aspect of this is: an untagged image is not very human readable and can be confusing to the user. A fedora image with tag `fedora:latest` is more readable than a none:none image with a 32 characters unique ImageID.

**Comment by tonistiigi** [comment]
> So in the above example if my image has N tags, I would need to do docker save tag1 tag2 tag3 tag4 ... tagN and create N tarballs of the same image to get my image and tags the same way they were at the source.

This is not how it works.  If you do `docker save tag1 tag2 tag3 tag4 ... tagN` you get a single tar that contains all the tags with no duplication of the actual layers.

**Comment by shishir-a412ed** [comment]
@tonistiigi Thanks for that info ! that works and solves my problem too. I would be more than happy to close this now.

**Comment by thaJeztah** [comment]
> This is not how it works. If you do docker save tag1 tag2 tag3 tag4 ... tagN you get a single tar that contains all the tags with no duplication of the actual layers.

Hm, actually didn't think of that. Good suggestion @tonistiigi 

@shishir-a412ed I'll go ahead and close this; happy to see it's resolved! :+1:

**Comment by aidanhs** [comment]
@tonistiigi it's more an interest in what the plans and figuring out whether they could help my use-cases (e.g. saving a single layer). If not, I would consider whether it would be appropriate to offer input etc.

**Pull request closed without merging**